File: Problems & Solutions/C/Complement/14 December 2022/Java/Complement.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    // Driver code
    public static void main(String[] args) throws Exception {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine().trim());
            String str = br.readLine().trim();

            Vector<Integer> ans = new Solve().findRange(str, n);
            if (ans.size() == 1) {
                System.out.println(ans.get(0));
            } else {
                System.out.println(ans.get(0) + " " + ans.get(1));
            }
        }
    }
}

// } Driver Code Ends

class Solve {

    Vector<Integer> findRange(String str, int n) {
        int max_so_far = Integer.MIN_VALUE, max_ending_here = 0, start = 0,
            end = 0, s = 0;
        boolean allOne = true;
        for (int i = 0; i < n; i++) {
            if (str.charAt(i) == '0') allOne = false;
        }
        Vector<Integer> ans = new Vector<Integer>();
        if (allOne) {
            ans.add(-1);
            return ans;
        }

        for (int i = 0; i < n; i++) {
            max_ending_here += (str.charAt(i) == '1' ? -1 : 1);

            if (max_so_far < max_ending_here) {
                max_so_far = max_ending_here;
                start = s;
                end = i;
            }

            if (max_ending_here < 0) {
                max_ending_here = 0;
                s = i + 1;
            }
        }
        ans.add(start + 1);
        ans.add(end + 1);
        return ans;
    }
}

File: Problems & Solutions/C/Complement/14 December 2022/C++/Complement.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;

// } Driver Code Ends
class Solution{
public:
    vector<int> findRange(string str, int n) {
        int max_so_far = INT_MIN, max_ending_here = 0, start = 0, end = 0, s = 0;
        bool allOne = true;
        for (int i = 0; i < n; i++) {
            if (str[i] == '0') allOne = false;
        }
    
        vector<int> ans;
        if (allOne) {
            ans.push_back(-1);
            return ans;
        }
        for (int i = 0; i < n; i++) {
            max_ending_here += (str[i] == '1' ? -1 : 1);
    
            if (max_so_far < max_ending_here) {
                max_so_far = max_ending_here;
                start = s;
                end = i;
            }
    
            if (max_ending_here < 0) {
                max_ending_here = 0;
                s = i + 1;
            }
        }
    
        ans.push_back(start + 1);
        ans.push_back(end + 1);
    
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        string s;
        cin >> s;
        Solution ob;
        auto ans = ob.findRange(s, n);

        if (ans.size() == 1) {
            cout << ans[0] << "\n";
        } else {
            cout << ans[0] << " " << ans[1] << "\n";
        }
    }
    return 0;
}
// } Driver Code Ends